### DISCUSSION TITLE
Gradle multiple string interpolation on one line not working

### DISCUSSION THREAD

**Discussion by siegenthalerroger** [renovatebot/renovate#18321] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

In our gradle files we use string interpolation to manage variables. This works very well when there is only one interpolation (`${xyz}`) on a line. As soon as there are multiple, things start breaking.

For example I have the following snippet in my root project:
```
subprojects {
    ext {
        apacheVersion = '3.12.0'
        apacheCompatibility = 'redhat'
        redhatVersion = '00001'
        apacheCommonsLangVersion = "${apacheVersion}.${apacheCompatibility}-${redhatVersion}"
```
and the following dependency definition in a subproject:
```
dependencies {
    implementation("org.apache.commons:commons-lang3:${apacheCommonsLangVersion}")
```

Furthermore I have configured a regex versioning matcher in renovate.json to actually be capable of finding updates despite the complicated version naming:
```
  "packageRules": [
    {
      "matchDatasources": [
        "maven"
      ],
      "matchPackageNames": ["org.apache.commons:commons-lang3"],
      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)\\.(?<compatibility>.*)-(?<build>.*)$"
    }
  ],
```

I'd expect renovate to detect the dependency and potentially inform of an applicable update (i.e. apacheVersion --> 3.13.0). Now I know that in the minimal reproduction linked below, there actually isn't a newer version of the package available. However renovate doesn't even detect that the dependency exists at all! Internally we have updates available that should match this so I'm certain that isn't the issue either. There aren't any Errors or anything in the logs so I sadly can't provide any further information.

**Minimal Reproduction:** https://github.com/siegenthalerroger/renovate-gradle-variable-interpolate-reproduction

Related to #18317 

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by Churro** [comment, marked as answer]
renovate doesn't perform interpolation on simple variable assignments. The package rule incl the custom versioning pattern you defined would only work if the dependency was extracted successfully.

Adding interpolation for constructs like `apacheCommonsLangVersion = "${apacheVersion}.${apacheCompatibility}-${redhatVersion}"` would be simple and that would make dependency extract work here. However, dependency updates would fail because the datasource wouldn't return `3.13.0` but `3.13.0.redhat-00001`. renovate wouldn't know how to split this string value and propagate it back to the original variables.

**Reply by siegenthalerroger** [reply]
The complexity with knowing how to update the original variables is very understandable but left me thinking about an alternative to our issue. Detecting an update is half the battle and another 40% of the way is informing developers that there's work to be done. If we could get it as far as opening an issue or PR to inform of an available update (without the actual required/working code-changes) that would already be 90% of the way there. Would this be a feature request that would have a chance?

**Reply by viceice** [reply]
i don't think so, as that doesn't fit to the renovate workflow

**Comment by viceice** [comment]
converted to discussion, as it's not a bug. also opening a feature request will probably never be done because of complexity.

**Comment by siegenthalerroger** [comment]
Ok I understand the reasoning but am left wondering how one would work with java dependencies that have explicit required "peers". Normally this is done by having the peer-dependency version in the version number (so something like the above but normally just `${dependencyVersion}-Dep${compatiblePeerDependencyVersion}`. 

Is there some other way to achieve this using an updateMatcher as well or some such?
